### PR TITLE
fix(auth): Keep redirect URL during 2FA setup and challenge

### PR DESCRIPTION
**Pull request by ChristophWurst**
## Summary

The redirect URL is crucial when a user is sent to Nextcloud to grant access for clients during the *login flow*. Without this, users will log in and land on the default page instead of the grant page for the setup.

## How to test

1. Log in as admin
2. Enforce 2FA
3. Install a 2FA provider like TOTP
4. Log out
5. Open /settings/user -> you get redirected to the login (note the redirect URL in the address bar)
6. Follow the steps to set up TOTP and confirm your OTP for the first time (always note the redirect URL in the address bar)

Master: land at /apps/dashboard, /apps/files or whatever the default is
Here: land at /settings/user

## TODO

- [x] ...

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: core/Middleware/TwoFactorMiddleware.php [hunk 1]
```diff
@@ -134,8 +134,10 @@ private function checkTwoFactor(Controller $controller, $methodName, IUser $user
 
 	public function afterException($controller, $methodName, Exception $exception) {
 		if ($exception instanceof TwoFactorAuthRequiredException) {
-			$params = [];
-			if (isset($this->request->server['REQUEST_URI'])) {
+			$params = [
+				'redirect_url' => $this->request->getParam('redirect_url'),
+			];
+			if (!isset($params['redirect_url']) && isset($this->request->server['REQUEST_URI'])) {
```
Comment (nickvergessen) [inline core/Middleware/TwoFactorMiddleware.php:140]: I still find it disturbing that isset is false when the value is null...

**Review by nickvergessen — APPROVED** [review]

**Review by st3iny — APPROVED** [review]
Tested and works.

**Pull request merged**